### PR TITLE
feat: add socks5 proxy support to kafka

### PR DESCRIPTION
**Pull request by alexole**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [ ] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by reimda** [comment]
Hi @alexole thanks for this work on socks5 proxy.  I see this is still a draft PR but I wanted to share some thoughts.

Since proxy support is something many other plugins could benefit from, I'd like to see us add it in a way that lets other plugins reuse it easily.  We have a similar situation with tls config.  In that case we have separate config structs with toml tags, and tls related code factored out of plugins.  This means every plugin that uses tls has settings with the same names and the plugin doesn't have to duplicate code to set up tls. (see plugins/common/tls/config.go)

I'd like to see us do something similar with proxy settings.  I expect you only need it for kafka but it would be helpful if you'd add it in a generic way.

**Comment by alexole** [comment]
@reimda thanks for the comment, I'll try to find time to make it properly.

**Comment by srebhan** [comment]
@alexole any news here?

**Comment by srebhan** [comment]
@alexole do you plan to continue on this PR?

**Comment by alexole** [comment]
Hi, sorry for not getting back to you. Yes, we will address the change requests soon.

File: plugins/outputs/kafka/kafka.go [hunk 1]
```diff
@@ -200,6 +204,12 @@ var sampleConfig = `
   ## Use TLS but skip chain & host verification
   # insecure_skip_verify = false
 
+  ## Optional SOCKS5 proxy to use when connecting to brokers
+  # enable_socks5 = true
```
[the remark this answers is not in the document]
Comment (mhoffm-aiven) [reply plugins/outputs/kafka/kafka.go]: renamed it to `socks5_enable`.

File: plugins/outputs/kafka/kafka.go [hunk 2]
```diff
@@ -320,6 +330,17 @@ func (k *Kafka) Connect() error {
 		}
 	}
 
+	if k.EnableSocks5 != nil && *k.EnableSocks5 {
+		config.Net.Proxy.Enable = true
+
+		dialer, err := k.Socks5ProxyConfig.GetDialer()
+		if err != nil {
+			log.Fatalf("Error while connecting to proxy server: %s", err)
+			return err
```
[the remark this answers is not in the document]
Comment (mhoffm-aiven) [reply plugins/outputs/kafka/kafka.go:310]: done

File: plugins/outputs/kafka/kafka.go [hunk 3]
```diff
@@ -39,6 +40,9 @@ type (
 		MaxRetry         int         `toml:"max_retry"`
 		MaxMessageBytes  int         `toml:"max_message_bytes"`
 
+		EnableSocks5 *bool `toml:"enable_socks5"`
```
[the remark this answers is not in the document]
Comment (mhoffm-aiven) [reply plugins/outputs/kafka/kafka.go]: Would it be ok to keep it? i like the explicit opt-in more

**Comment by mhoffm-aiven** [comment]
Hey @srebhan,

could you have another look?

File: plugins/outputs/kafka/kafka.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/outputs/kafka/kafka.go]: Fine with me.

File: plugins/outputs/kafka/kafka.go [hunk 4]
```diff
@@ -292,6 +301,16 @@ func (k *Kafka) Init() error {
 		k.TLSKey = k.Key
 	}
 
+	if k.Socks5ProxyEnabled {
+		config.Net.Proxy.Enable = true
+
+		dialer, err := k.Socks5ProxyConfig.GetDialer()
+		if err != nil {
+			return fmt.Errorf("Error while connecting to proxy server: %s", err)
```
Comment (srebhan) [inline plugins/outputs/kafka/kafka.go:309]: In golang error messages should start with a lowercase letter.
```suggestion
			return fmt.Errorf("connecting to proxy server failed: %s", err)
```

File: plugins/common/proxy/socks5_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/armon/go-socks5"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSocks5ProxyConfig(t *testing.T) {
```
Comment (srebhan) [inline plugins/common/proxy/socks5_test.go:12]: You need to guard this test to be excluded in short mode as there is no guarantee of a functioning network (beside localhost)...
```suggestion
func TestSocks5ProxyConfig(t *testing.T) {
        if testing.Short() {
                t.Skip("Skipping integration test in short mode")
        }

```

File: plugins/common/proxy/socks5_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	require.NoError(t, err)
+
+	var proxyConn net.Conn
+	for i := 0; i < 10; i++ {
+		proxyConn, err = dialer.Dial("tcp", l.Addr().String())
+		if err != nil {
+			time.Sleep(10 * time.Millisecond)
+			continue
+		}
+		break
```
Comment (srebhan) [inline plugins/common/proxy/socks5_test.go:49]: Can we please beautify this
```suggestion
		if err == nil {
			break
		}
		time.Sleep(10 * time.Millisecond)
```

**Review by srebhan — COMMENTED** [review]
Sorry @alexole for the late response. I have some smaller comments left, but nothing big. Can you please take a look?

File: plugins/outputs/kafka/kafka.go (hunk 4 above)
Comment (mhoffm-aiven) [reply plugins/outputs/kafka/kafka.go]: Looks good, i commited it straight from github.

File: plugins/common/proxy/socks5_test.go (hunk 5 above)
Comment (mhoffm-aiven) [reply plugins/common/proxy/socks5_test.go:12]: sounds reasonable, commited it straight from github

File: plugins/common/proxy/socks5_test.go (hunk 6 above)
Comment (mhoffm-aiven) [reply plugins/common/proxy/socks5_test.go]: Looks good, committed straight from github.

**Comment by srebhan** [comment]
Sorry @mhoffm-aiven I messed up the formatting. Can you please run `make fmt` and push the change... :-(

**Comment by srebhan** [comment]
Oh... too slow. ;-)

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for pushing this over the finishing line @mhoffm-aiven!

File: plugins/outputs/kafka/kafka.go [hunk 7]
```diff
@@ -292,6 +301,16 @@ func (k *Kafka) Init() error {
 		k.TLSKey = k.Key
 	}
 
+	if k.Socks5ProxyEnabled {
```
Comment (powersj) [inline plugins/outputs/kafka/kafka.go:304]: How does this get set by an end-user?

Comment (mhoffm-aiven) [reply plugins/outputs/kafka/kafka.go:304]: That can be configured by writing `socks5_enabled: true` in the plugin configuration

```
type Socks5ProxyConfig struct {
	Socks5ProxyEnabled  bool   `toml:"socks5_enabled"`
	Socks5ProxyAddress  string `toml:"socks5_address"`
	Socks5ProxyUsername string `toml:"socks5_username"`
	Socks5ProxyPassword string `toml:"socks5_password"`
}
```

Comment (powersj) [reply plugins/outputs/kafka/kafka.go:304]: Ah right, I see it in the sample config above, can you please add it to the README as well?

Comment (mhoffm-aiven) [reply plugins/outputs/kafka/kafka.go:304]: Ah, forgot that! Added it just now.

Comment (powersj) [reply plugins/outputs/kafka/kafka.go:304]: thank you!

**Review by powersj — APPROVED** [review]

**Pull request merged**